lib/sycamore/sycamore/transforms/map.py

```diff
@@ -62,20 +63,39 @@ def custom_flat_mapping_function(document: Document) -> list[Document]:
     """
 
     def __init__(self, child: Node, *, f: Callable[[Document], list[Document]], **resource_args):
-        super().__init__(child, **resource_args)
-        self._f = f
-
-    def execute(self) -> "Dataset":
-        input_dataset = self.child().execute()
-        if isinstance(self._f, type):
-            ray_callable = generate_flat_map_class(self._f)
-            return input_dataset.flat_map(ray_callable, compute=ActorPoolStrategy(size=1), **self.resource_args)
+        super().__init__(child, f=FlatMap.wrap(f), name=f.__name__, **resource_args)
+
+    @staticmethod
+    def wrap(f: Callable[[Document], list[Document]]) -> Callable[[list[Document]], list[Document]]:
+        if isinstance(f, type):
+
+            class _Wrap(f):  # type: ignore[valid-type,misc]
+                def __init__(self, *args, **kwargs):
+                    super().__init__(*args, **kwargs)
+
+                def __call__(self, docs):
+                    s = super()
+                    ret = []
+                    for d in docs:
+                        ret.extend(s.__call__(d))
+                    return ret
+
+            return _Wrap
         else:
-            ray_callable = generate_flat_map_function(self._f)
-            return input_dataset.flat_map(ray_callable, **self.resource_args)
 
+            def _wrap(docs):
+                assert isinstance(docs, list)
+                ret = []
+                for d in docs:
+                    assert isinstance(d, Document)
```

Comment:
should these asserts also be in the class part?

Reply:
They were for helping me debug problems, but don't hurt to put in both branches, so I did.